modules/core/nuxeo-core-bulk/src/main/java/org/nuxeo/ecm/core/bulk/computation/BulkScrollerComputation.java

```diff
@@ -206,12 +208,16 @@ protected void processRecord(ComputationContext context, Record record) {
                 updateStatusAfterScroll(context, commandId, documentCount, limitReached);
             }
         } catch (IllegalArgumentException | QueryParseException | DocumentNotFoundException e) {
+            int statusCode = HttpStatus.SC_INTERNAL_SERVER_ERROR;
+            if (e instanceof NuxeoException) {
+                statusCode = ((NuxeoException) e).getStatusCode();
+            }
```

Comment:
I think we want to turn `IllegalArgumentException` into a 400 as well.
For the `DocumentNotFoundException` I think there's a discussion to have to decide if we want to keep a 404 or turn it into a 400. I don't know when this exception is generated.

Reply:
yes, agree on that. But I told myself if we find the issue and we align it to the same error code. It can break changes. But I'll take a look.

Reply:
Currently, the exception is raised [Here](https://github.com/nuxeo/nuxeo/blob/master/modules/core/nuxeo-core-storage-sql/nuxeo-core-storage-sql/src/main/java/org/nuxeo/ecm/core/storage/sql/SessionImpl.java#L576). I did a change on this one to make it more consistent with DBS. But I think we have to discuss the others ones, where we raise `IllegalArgumentException` for example [Here](https://github.com/nuxeo/nuxeo/blob/master/modules/core/nuxeo-core-storage-sql/nuxeo-core-storage-sql/src/main/java/org/nuxeo/ecm/core/storage/sql/SessionImpl.java#L584)

Reply:
Hm changing code in `SessionImpl` is going too far for this commit. And there'd be many other similar changes to do if we want to be consistent, and probably unit tests to update etc.
Here you should just in the `catch` clause check if we have an `IllegalArgumentException` and then set the status to SC_BAD_REQUEST.

Reply:
ah sorry, I see your point. I though we need to do it a low level.

Reply:
My " I don't know when this exception is generated" was about the `DocumentNotFoundException` as I don't know if it's an internal one that we catch for unrelated reason (in which case 500 is ok), or if it's a direct consequence of the user passing a reference to a doc that doesn't exist (in which case it's a 400).

Reply:
yep, 
 For the query `SELECT * FROM Document WHERE ecm:path = 'non/existing/path'` we have two different logic: 

- DBS low level is raising a `DocumentNotFoundException` which is `404` [DBS#404](https://github.com/nuxeo/nuxeo/blob/master/modules/core/nuxeo-core-storage-dbs/src/main/java/org/nuxeo/ecm/core/storage/dbs/DBSSession.java#L266)
- VCS on other side is raising a `IllegalArgumentException` which is [converted to 500](https://github.com/nuxeo/nuxeo/pull/4606/files#diff-1723bb85f2e3ac2876b96b87404ede9ffee1f4d9747e576e0cfdb80bfee5b1bbR170) by myself. That's why I pointed you initially to the `SessionImpl` and I did the changes at its level. which I agree has more impact. 

Perhaps, I am wrong, but for me we have to make more consistency on low level VSC vs DBS and the raising exception. Otherwise we can rework [Here](https://github.com/nuxeo/nuxeo/pull/4606/files#diff-1723bb85f2e3ac2876b96b87404ede9ffee1f4d9747e576e0cfdb80bfee5b1bbR170-R173) to turn into the one(s) that we want:

- `IllegalArgumentException` into `DocumentNotFoundException` => both of them will be `404`
- We consider `IllegalArgumentException` and `DocumentNotFoundException` as `400`
 
for me the two options make sens as `ecm:path = 'non/existing/path'` can be considered as not exiting or a bad request.

Reply:
Ok but fixing all this is too much for this PR which is about error handling in the async adapter. Here turning `IllegalArgumentException` into a 400 in the catch clause is enough and makes sense semantically. We'll have 400 vs 404 for VCS/DBS but I think it's good enough for a client. Unifying this at low level is another topic.

Reply:
I made more tests and it seems that for DBS (mongo) if we change the query to `nxql = "SELECT * FROM Document WHERE ecm:path = '/non/existing/path'";` I mean adding `/` at the beginning of `ecm:path` predicate. We don't have the error at all, because it returns `null` [Here](https://github.com/nuxeo/nuxeo/blob/master/modules/core/nuxeo-core-storage-dbs/src/main/java/org/nuxeo/ecm/core/storage/dbs/DBSSession.java#L328)

Reply:
sorry, perhaps I am missing things. But https://github.com/nuxeo/nuxeo/pull/4606#discussion_r598933252 will not be enough, we have to make consistency between both of `IllegalArgumentException` and `DocumentNotFoundException ` as DBS is returning `DocumentNotFoundException`.  Otherwise my test will not pass. If want to remove the original check that I have added. Can I convert `IllegalArgumentException` to 404. Will make consistency between the two types (VCS and DBS).

Reply:
Turning `IllegalArgumentException` to 404 is wrong. However we could very well convert `DocumentNotFoundException` to a 400 here in this `catch` block, I think the semantics still work. @nuxeo/platform any comment?

Reply:
It should be fine as it seems to be an edge case for me.